packages/convex-helpers/server/triggers.ts

```diff
@@ -0,0 +1,267 @@
+import { GenericId } from "convex/values";
+import { Mod } from "./customFunctions.js";
+
+/**
+ * This function will be called when a document in the table changes.
+ */
+export type Trigger<
+  Ctx,
+  DataModel extends GenericDataModel,
+  TableName extends TableNamesInDataModel<DataModel>,
+> = (
+  ctx: Ctx & { innerDb: GenericDatabaseWriter<DataModel> },
+  change: Change<DataModel, TableName>,
+) => Promise<void>;
+
+export type Change<
+  DataModel extends GenericDataModel,
+  TableName extends TableNamesInDataModel<DataModel>,
+> = {
+  id: GenericId<TableName>;
+} & ({
+  operation: "insert";
+  oldDoc: null
+  newDoc: DocumentByName<DataModel, TableName>;
+} | {
+  operation: "update";
+  oldDoc: DocumentByName<DataModel, TableName>;
+  newDoc: DocumentByName<DataModel, TableName>;
+} | {
+  operation: "delete";
+  oldDoc: DocumentByName<DataModel, TableName>;
+  newDoc: null;
+});
+
+/**
+ * Construct Triggers to register functions that run whenever a table changes.
+ * Sample usage:
+ * 
+ * ```
+ * import { mutation as rawMutation } from "./_generated/server";
+ * import { DataModel } from "./_generated/dataModel";
+ * import { Triggers } from "convex-helpers/server/triggers";
+ * 
+ * const triggers = new Triggers<DataModel>();
+ * triggers.register("myTableName", async (ctx, change) => {
+ *   console.log("Table changed", change);
+ * });
+ * // Use `mutation` to define all mutations, and the triggers will get called.
+ * export const mutation = customMutation(rawMutation, triggers.customFunctionWrapper());
+ * ```
+ */
+export class Triggers<
+  DataModel extends GenericDataModel,
+  Ctx extends { db: GenericDatabaseWriter<DataModel> } = GenericMutationCtx<DataModel>,
+> {
+  registered: {
+    [TableName in TableNamesInDataModel<DataModel>]?: Trigger<Ctx, DataModel, TableName>[];
+  } = {};
+  
+  register<TableName extends TableNamesInDataModel<DataModel>>(
+    tableName: TableName,
+    trigger: Trigger<Ctx, DataModel, TableName>,
+  ) {
+    if (!this.registered[tableName]) {
+      this.registered[tableName] = [];
+    }
+    this.registered[tableName]!.push(trigger);
+  }
+
+  customFunctionWrapper(): Mod<Ctx, {}, {innerDb: GenericDatabaseWriter<DataModel>}, {}> {
+    const triggers = this;
+    return {
+      args: {},
+      input(ctx: Ctx) {
+        const innerDb = ctx.db;
+        return { ctx: {
+          db: new DatabaseWriterWithTriggers(ctx, innerDb, triggers),
+          innerDb,
+        }, args: {} };
+      },
+    };
+  }
+}
+
+class Lock {
+  promise: Promise<void> | null = null;
+  resolve: (() => void) | null = null;
+
+  async lock(): Promise<(() => void)> {
+    while (this.promise !== null) {
+      await this.promise;
+    }
+    [this.promise, this.resolve] = this._newLock();
+    return () => {
+      this.promise = null;
+      this.resolve?.();
+    };
+  }
```

Comment:
would it make sense to do a `withLock` API so it's much harder to forget to unlock?